### PR TITLE
Refactor compositor and modifier loading and structure

### PR DESCRIPTION
**Pull request by djhoese**
This is a major refactor of the compositor and modifier system in Satpy. This PR:

1. Creates a `ModifierBase` class for all modifiers to be based off of.
2. Moves all existing modifier classes to a new `satpy/modifiers/` sub-package.
3. Reorganizes the compositor and modifier loading to be more understandable (hopefully).
4. Deprecates modifiers using `compositor: ` key to point to the python class to use and instead favor `modifier: `.
5. Fixes small things that may cause issues in the future. For example, I found some uses of dictionary keys like:

   ```
   for sensor in sensor_dict.keys():
       if sensor in X: return X[sensor]
   ```

   Code like this could produce non-deterministic results as it will return the *first* result found. I've added a `sorted` to the sensor keys usage so this should at least always pass or always fail.

TODO:

 - [ ] Split `satpy/modifiers/__init__.py` up even more (atmosphere.py, etc). I'm not sure the best names for these as I'm not sure how similar some of the modifiers are to each other (like are CO2, rayleigh, and sunz all technically similar?).
 - [x] Possibly move the modifier config parsing and compositor config parsing to helper classes (currently discussing on slack)
 - [x] Move compositor loader class(es) to a separate submodule of `satpy/composites/` and out of `__init__.py`.

 - [x] Passes ``flake8 satpy`` 

I'm making this a regular PR (not draft) to invite reviews now.

### CODE REVIEW

File: satpy/dependency_tree.py [hunk 1]
```diff
@@ -441,7 +441,7 @@ def _promote_query_to_modified_dataid(self, query, dep_key):
 
     def get_compositor(self, key):
         """Get a compositor."""
-        for sensor_name in self.compositors.keys():
+        for sensor_name in sorted(self.compositors.keys()):
```
Comment (mraspaud) [inline satpy/dependency_tree.py:444]: Why the sorting? Is it tested?

File: satpy/dependency_tree.py [hunk 2]
```diff
@@ -453,7 +453,7 @@ def get_modifier(self, comp_id):
         """Get a modifer."""
         # create a DataID for the compositor we are generating
         modifier = comp_id['modifiers'][-1]
-        for sensor_name in self.modifiers.keys():
+        for sensor_name in sorted(self.modifiers.keys()):
```
Comment (mraspaud) [inline satpy/dependency_tree.py:456]: Why the sorting? Is it tested?

**Review by mraspaud — COMMENTED** [review]
Will look more later, but from a first look, impressive work!

File: satpy/dependency_tree.py (hunk 1 above)
Comment (pnuu) [inline satpy/dependency_tree.py:444]: `.keys()` isn't needed.

File: satpy/dependency_tree.py (hunk 2 above)
Comment (pnuu) [inline satpy/dependency_tree.py:456]: Same here, `.keys()` isn't needed.

**Review by pnuu — COMMENTED** [review]
Looks good! Just couple comments.

File: satpy/dependency_tree.py (hunk 1 above)
Comment (djhoese) [reply satpy/dependency_tree.py:444]: See my description at the top. It isn't tested as far as I know and is a lightly used feature at best (supporting multiple sensors in a single Scene). I should add a test.

Comment (djhoese) [reply satpy/dependency_tree.py]: Fixed. (wasn't my original code anyway)

File: satpy/dependency_tree.py (hunk 2 above)
Comment (djhoese) [reply satpy/dependency_tree.py]: Fixed.

**Review by pnuu — APPROVED** [review]
Apart from the (new) merge conflicts and the Stickler complaints, the PR looks good. I think the current split to different categories is good and can be finetuned later.

I think the `compositor:` items in the composite YAML files ought to be changed to `modifiers:` to show the preferred/proper usage and to silence the unnecessary deprecation warnings caused by the built-in composites.

**Comment by djhoese** [comment]
@pnuu 

> I think the compositor: items in the composite YAML files ought to be changed to modifiers: to show the preferred/proper usage and to silence the unnecessary deprecation warnings caused by the built-in composites.

I already did that. I'm confused? Which ones are you talking about.

**Comment by pnuu** [comment]
> > I think the compositor: items in the composite YAML files ought to be changed to modifiers: to show the preferred/proper usage and to silence the unnecessary deprecation warnings caused by the built-in composites.
> 
> I already did that. I'm confused? Which ones are you talking about.

Oh, sorry! The `Files changed` page hid these changes. After adjusting the filters to hide Python files they were visible. All good then :smile:

**Comment by djhoese** [comment]
deepcode apparently ignores the `# noqa` comments. Too bad.

File: satpy/modifiers/__init__.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+#
+# satpy is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# satpy.  If not, see <http://www.gnu.org/licenses/>.
+"""Modifier classes and other related utilities."""
+
+from .base import ModifierBase  # noqa
```
Comment (pnuu) [inline satpy/modifiers/__init__.py:20]:
```suggestion
from .base import ModifierBase  # noqa: F401
```

File: satpy/modifiers/__init__.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+from .spectral import NIREmissivePartFromReflectance, NIRReflectance  # noqa
```
Comment (pnuu) [inline satpy/modifiers/__init__.py:21]:
```suggestion
from .spectral import NIREmissivePartFromReflectance, NIRReflectance  # noqa: F401
```

File: satpy/modifiers/__init__.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+from .geometry import SunZenithCorrector, EffectiveSolarPathLengthCorrector  # noqa
```
Comment (pnuu) [inline satpy/modifiers/__init__.py:22]:
```suggestion
from .geometry import SunZenithCorrector, EffectiveSolarPathLengthCorrector  # noqa: F401
```

File: satpy/modifiers/__init__.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+from .atmosphere import PSPRayleighReflectance, PSPAtmosphericalCorrection, CO2Corrector  # noqa
```
Comment (pnuu) [inline satpy/modifiers/__init__.py:23]:
```suggestion
from .atmosphere import PSPRayleighReflectance, PSPAtmosphericalCorrection, CO2Corrector  # noqa: F401
```

**Review by pnuu — COMMENTED** [review]
DeepCode doesn't like `# noqa` as catch-all, specific skips should work. At least with my Pyresample PR `# noqa: F401` worked, so I'll commit them and we'll see how that goes.

File: satpy/modifiers/__init__.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+#
+# satpy is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# satpy.  If not, see <http://www.gnu.org/licenses/>.
+"""Modifier classes and other related utilities."""
+
+from .base import ModifierBase  # noqa: F401
```
Comment (pnuu) [inline satpy/modifiers/__init__.py:20]:
```suggestion
# file deepcode ignore W0611: Ignore unused imports in init module

from .base import ModifierBase  # noqa: F401
```

**Comment by pnuu** [comment]
Ok, that worked. The DeepCode `Details` page had a button that showed how to ignore the warnings.

**Comment by djhoese** [comment]
Hm I wonder why the individual line comments didn't work. I was hoping to avoid the whole module thing but I guess as long as it works for now.

**Comment by pnuu** [comment]
The other option would be to place `# deepcode ignore W0611: <reason here>` above every offending line. I think this "global" one is better, as we just suppress one type of warning we'd need to use for every import.

**Pull request merged**